tempesta_fw/sock.c

```diff
@@ -1162,17 +1275,126 @@ ss_tx_action(void)
 		default:
 			BUG();
 		}
+dead_sock:
 		sock_put(sk); /* paired with tfw_wq_push() */
 	}
 }
 
+#define HANDLE_TOO_LONG_WAIT(t0, job)					\
+do {									\
+	if (t0 + HZ * 5 < jiffies) {					\
+		SS_WARN("pending " job " for 5s (connections count %#lx)\n",\
+			acc);						\
+		/*							\
+		 * Something is very wrong with socket accounting.	\
+		 * Give up waiting and hope for the best.		\
+		 */							\
+		return;							\
+	}								\
+} while (0)
+
+void
+ss_wait_listeners(void)
+{
+	int cpu;
+	long acc = 0, acc_old = 0;
+	unsigned long t0 = jiffies;
+
+	might_sleep();
+	while (1) {
+		for_each_online_cpu(cpu)
+			acc += atomic64_read(per_cpu_ptr(&__ss_act_cnt, cpu));
+		BUG_ON(acc < 0);
+		if (!(acc & SS_M_ACT_NEWCONN))
+			break;
+		schedule();
+		if (acc == acc_old) {
+			HANDLE_TOO_LONG_WAIT(t0, "listeners");
```

Comment:
Replaced the message string by "listening sockets". Actually this phase synchronize listening sockets, not just new connections.